container-conf/playwright/tests/testing-utils.js

```diff
@@ -0,0 +1,279 @@
+    }
+
+    let matchedHostname = matches[0][1];
+    return link.replace(matchedHostname, newHost);
+}
+
+async function getUserHome() {
+    let homePromise = new Promise((resolve, reject) => {
+        exec('echo ~',
+            function (error, stdout, stderr) {
+                if (error !== null) {
+                    reject(error)
+                }
+                resolve(stdout.trim());
+            });
+    })
+
+    return await homePromise;
+}
+
+export let loginIfNeeded = async (page, applicationName) => {
+    if(NEEDS_EMAIL_LOGIN) {
+        await loginWithEmail(page, applicationName);
+    } else {
+        await loginAsGuest(page, applicationName);
+    }
+}
+
+export let loginAsGuest = async (page, applicationName) => {
+    await page.goto(HOST + "/" + applicationName + "#/login-with/guest");
+}
+
+export let loginWithEmail = async (page, applicationName, email="vagrant@localhost.localdomain") => {
+    let mailManager = new MailManager((await getUserHome()) + "/mail");
+    mailManager.deleteAllEmails();
+
+    await page.goto(HOST + "/" + applicationName + "#/login-with/email");
+    let emailFormGroup = page.locator(".form-group", { has: page.locator(textFuzzyEquals("Your Email")) });
+    await page.waitForTimeout(1000);
+    await emailFormGroup.locator("input").type(email);
+    await page.waitForTimeout(1000);
+    await page.locator("button").locator(textFuzzyEquals("Continue")).click();
+    
+    //let link = replaceHostname((await mailManager.getFirstEmailLink(500, 30)), HOST);
+    let link = await mailManager.getFirstEmailLink(500, 30);
+    //console.log("mail before delete: " + JSON.stringify(mailManager.listMail()))
+    mailManager.deleteAllEmails();
+    //console.log("mail after delete: " + JSON.stringify(mailManager.listMail()))
+    
+    console.log(link);
+    await page.goto(link);
+
+    await page.waitForTimeout(250);
+    try {
+        let continueButton = page.locator("button", { has: page.locator(textFuzzyEquals("Confirm")) });
+        await continueButton.waitFor({ state: 'attached', timeout: 2500 });
+        await continueButton.click();
+    } catch {
+        await page.locator("input[name=displayName]").type("Test"); // TODO unique name
+        await page.waitForTimeout(250);
+        await page.locator("button", { has: page.locator(textFuzzyEquals("Submit")) }).click();
+    }
+    await page.waitForURL(`${HOST}/${applicationName}#/simulations`)
+}
+
+export let navigateToSimulation = async (page, simFolderNames) => {
+    for(var path of simFolderNames) {
+        //await page.waitForTimeout(1000);
+        await page.locator("div.sr-thumbnail span")
+        .locator(textFuzzyEquals(path)).click();
+        await page.waitForTimeout(500);
+    }
+}
+
+export let navigateToFirstSimulation = async (page) => {
+    let enterNextItem = async (page) => {
+        let firstItemLocator = page.locator(".sr-icon-col div.sr-thumbnail", { has: page.locator("span.glyphicon") }).first();
```

Comment:
I find the names too long. Naming a variable elementHandle which is only used on the next line means I have to think about about "what is elementHandle?" If it is x or something like that, it's just a temporary, and the name is uninteresting.